gtp2ogs.js

```diff
@@ -169,11 +180,160 @@ class Bot {
+        //
+        if (NOCLOCK) return;
+
+        let black_offset = 0;
+        let white_offset = 0;
+
+        let now = state.clock.now ? state.clock.now : Date.now();
+
+        if (state.clock.current_player == state.clock.black_player_id) {
+            black_offset = (now - state.clock.last_move) / 1000;
+        } else {
+            white_offset = (now - state.clock.last_move) / 1000;
+        }
+
+        if (state.time_control.system == 'byoyomi') {
+            // GTP spec says time_left should have 0 for stones until main_time has run out.
+            //
+            // If the bot connects in the middle of a byoyomi period, it won't know how much time it has left before the period expires.
+            // When restarting the bot mid-match during testing, it sometimes lost on timeout because of this. To work around it, we can
+            // reduce the byoyomi period size by the offset. Not strictly accurate but GTP protocol provides nothing better. Once bot moves
+            // again, the next state setup should have this corrected. This problem would happen if a bot were to crash and re-start during
+            // a period. This is only an issue if it is our turn, and our main time left is 0.
+            //
+            // TODO: If I add support for a persistant bot connection (not restarting process each move), be sure to think about mid-match
+            // reconnect time settings in more depth.
+            //
+            if (KGSTIME) {
+                let black_timeleft = Math.max( Math.floor(state.clock.black_time.thinking_time - black_offset), 0);
+                let white_timeleft = Math.max( Math.floor(state.clock.white_time.thinking_time - white_offset), 0);
+
+                this.command("kgs-time_settings byoyomi " + state.time_control.main_time + " "
+                    + Math.floor(state.time_control.period_time -
+                        (state.clock.current_player == state.clock.black_player_id ? black_offset : white_offset)
+                    )
+                    + " " + state.time_control.periods);
+                this.command("time_left black " + black_timeleft + " " + (black_timeleft > 0 ? "0" : state.clock.black_time.periods));
+                this.command("time_left white " + white_timeleft + " " + (white_timeleft > 0 ? "0" : state.clock.white_time.periods));
+            } else {
+                // OGS enforces the number of periods is always 1 or greater. Let's pretend the final period is a Canadian Byoyomi of 1 stone.
+                // This lets the bot know it can use the full period per move, not try to fit the rest of the game into the time left.
+                //
+                let black_timeleft = Math.max( Math.floor(state.clock.black_time.thinking_time
+                    - black_offset + (state.clock.black_time.periods - 1) * state.time_control.periods), 0);
```

Comment:
Shouldn't this be `(state.clock.black_time.periods - 1) * state.time_control.periods_time)` ?